### PR TITLE
Clean up how `generate-at-breakpoints` work

### LINKED ISSUE #108: Remove all the Sass for the `generate-at-breakpoints` mixin as it runs the risk of a large compiled style.css file
Discussed on HipChat:

> > Chris Pearce
> > 10:08
> > @here worried about the `generate-at-breakpoints` mixin causing monster CSS files. I'm thinking remove it entirely from Scally, like keep the mixin but that's it then it's up to the fw author to create them when needed maybe in style.scss e.g.
> > Chris Pearce
> > 10:08
> > @import "utilities/u-alignments";
> > @include generate-at-breakpoints('.u-align-v-inline-top',
> >   palm) {
> >   vertical-align: top;
> > }
> > Chris Pearce
> > 10:08
> > @import "utilities/u-alignments";
> > @include generate-at-breakpoints('.u-align-v-inline-top',
> >   palm) {
> >   vertical-align: top;
> > }
> > Chris Pearce
> > 10:08
> > anyway discuss at 11am
> > Chris Pearce
> > 10:13
> > rather than this:
> > Chris Pearce
> > 10:13
> > $u-alignments-breakpoints: palm;
> > @import "utilities/u-alignments";
> > Chris Pearce
> > 10:13
> > which outputs this:
> > Chris Pearce
> > 10:14
> > .u-align-v-inline-top {
> >   vertical-align: top; }
> > @media (max-width: 40em) {
> >   .u-align-v-inline-top-up-to-palm {
> >     vertical-align: top; } }
> > .u-align-v-inline-middle {
> >   vertical-align: middle; }
> > @media (max-width: 40em) {
> >   .u-align-v-inline-middle-up-to-palm {
> >     vertical-align: middle; } }
> > .u-align-v-inline-baseline {
> >   vertical-align: baseline; }
> > @media (max-width: 40em) {
> >   .u-align-v-inline-baseline-up-to-palm {
> >     vertical-align: baseline; } }
> > .u-align-v-inline-bottom {
> >   vertical-align: bottom; }
> > @media (max-width: 40em) {
> >   .u-align-v-inline-bottom-up-to-palm {
> >     vertical-align: bottom; } }
> > .u-align-v-inline-inherit {
> >   vertical-align: inherit; }
> > @media (max-width: 40em) {
> >   .u-align-v-inline-inherit-up-to-palm {
> >     vertical-align: inherit; } }
> > /**
> > - Block level vertical alignments.
> >   _/
> >   .u-align-v-block-middle-flexbox {
> >     display: flex;
> >     align-items: center;
> >     justify-content: center; }
> >   @media (max-width: 40em) {
> >     .u-align-v-block-middle-flexbox-up-to-palm {
> >       display: flex;
> >       align-items: center;
> >       justify-content: center; } }
> >   .u-align-v-block-middle-table {
> >     display: table;
> >     height: 100%;
> >     width: 100%; }
> >   @media (max-width: 40em) {
> >     .u-align-v-block-middle-table-up-to-palm {
> >       display: table;
> >       height: 100%;
> >       width: 100%; } }
> >   .u-align-v-block-middle-table__inner {
> >     display: table-cell;
> >     vertical-align: middle; }
> >   @media (max-width: 40em) {
> >     .u-align-v-block-middle-table__inner-up-to-palm {
> >       display: table-cell;
> >       vertical-align: middle; } }
> >   .u-align-v-block-middle-position {
> >     position: relative;
> >     top: 50%;
> >     transform: translateY(-50%); }
> >   @media (max-width: 40em) {
> >     .u-align-v-block-middle-position-up-to-palm {
> >       position: relative;
> >       top: 50%;
> >       transform: translateY(-50%); } }
> >   /_*
> > - Center align a block level element.
> >   *
> > - 1. Make it work on inline elements e.g. `img`.
> >   */
> >   .u-align-center-block {
> >     display: block;
> >     margin-left: auto;
> >     margin-right: auto; }
> >   @media (max-width: 40em) {
> >     .u-align-center-block-up-to-palm {
> >       display: block;
> >       margin-left: auto;
> >       margin-right: auto; } }
> >   Hide full text
> >   Alec Raeside
> >   10:15
> >   yea it doesnt seem worth it
> >   Chris Pearce
> >   10:15
> >   it's too dangerous
> >   Chris Pearce
> >   10:15
> >   just means its a big non-backwards compatible change to Scally

### PR DESCRIPTION
**Pull request by chris-pearce**
This closes #108.

### CODE REVIEW

File: utilities/_u-display.scss [hunk 1]
```diff
@@ -19,68 +19,99 @@
...
 
+// Block
 %u-display-block,
 .u-display-block {display: block;}
 
-@include generate-at-breakpoints('.u-display-block', $u-display-breakpoints) {
-  display: block;
-}
+@if $u-display-breakpoint-toggle-block {
+  @include generate-at-breakpoints('.u-display-block', $u-display-breakpoints) {display: block;}
```
Comment (smgcode) [inline utilities/_u-display.scss:65]: Should these lines be less than 80 characters?

File: utilities/_u-drop-down.scss [hunk 2]
```diff
@@ -92,6 +107,8 @@ $u-drop-down-touch-support-class:   is-touch !default;
   .#{$u-drop-down-touch-support-class} .u-drop-down--on-hover .u-drop-down__target
   {display: none;}
 
-  @include generate-at-breakpoints('.#{$u-drop-down-touch-support-class} .u-drop-down--on-hover .u-drop-down__target', $u-drop-down-breakpoints) {
-    display: none;
-  }
\ No newline at end of file
+  @if $u-drop-down-breakpoint-toggle-on-hover {
+    @include generate-at-breakpoints('.#{$u-drop-down-touch-support-class} .u-drop-down--on-hover .u-drop-down__target', $u-drop-down-breakpoints) {
```
Comment (smgcode) [inline utilities/_u-drop-down.scss:111]: 80 characters here too? Will running lint on this catch things like this?

Comment (chris-pearce) [inline utilities/_u-drop-down.scss:111]: @AlecRaeside ^

Comment (chris-pearce) [inline utilities/_u-drop-down.scss:111]: https://github.com/westfieldlabs/scally/issues/116

**Comment by AlecRaeside** [comment]
LGTM

**Comment by chris-pearce** [comment]
@AlecRaeside @kllevin @smgcode when this goes in and you get latest of Scally you'll have to refactor your `style.scss` file to accommodate this new setup.

**Comment by AlecRaeside** [comment]
LGTM

**Comment by AlecRaeside** [comment]
So to confirm, when this goes in we'll review all the `from` and `up-to` classes we're using and turn their variables on?

**Comment by chris-pearce** [comment]
@AlecRaeside yeah so wherever you've declared `-breakpoints` settings in `style.scss` e.g. https://github.com/westfield/aem_centre-site/blob/master/content/src/main/content/jcr_root/etc/designs/centre-site/style/scss/style.scss#L217 you'll need to update it to match the new setup.

File: layout/_l-grid-pull.scss [hunk 3]
```diff
@@ -20,10 +20,10 @@
  */
 
 // Turned off by default as this generates a lot of CSS
-$l-grid-pull: false !default;
+$l-grid-pull:               true !default;
```
Comment (kllevin) [inline layout/_l-grid-pull.scss]: Why is this set to true? It seems like all the new toggles are set to false by default, should we set this up to work the same way?

File: layout/_l-grid-push.scss [hunk 4]
```diff
@@ -20,10 +20,10 @@
  */
 
 // Turned off by default as this generates a lot of CSS
-$l-grid-push: false !default;
+$l-grid-push:               true !default;
```
Comment (kllevin) [inline layout/_l-grid-push.scss]: Same as `pull`

File: utilities/_u-alignments.scss [hunk 5]
```diff
@@ -17,8 +17,33 @@
...
+
+$u-align-breakpoint-toggle-v-inline-bottom:              false !default;
+
+$u-align-breakpoint-toggle-v-inline-inherit:             false !default;
+
+$u-align-breakpoint-toggle-v-block-middle-flexbox:       false !default;
+
+$u-align-breakpoint-toggle-v-block-middle-table:         false !default;
+
+$u-align-breakpoint-toggle-v-block-middle-table__inner:  false !default;
```
Comment (kllevin) [inline utilities/_u-alignments.scss]: Why is this toggle name formatted differently should it be `$u-align-breakpoint-toggle-v-block-middle-table-inner` instead?

File: utilities/_u-alignments.scss [hunk 6]
```diff
@@ -109,11 +153,13 @@ $u-alignments-breakpoints: $default-breakpoints !default;
     vertical-align: middle;
   }
 
-  @include generate-at-breakpoints('.u-align-v-block-middle-table__inner',
-    $u-alignments-breakpoints) {
-    display: table-cell;
-    vertical-align: middle;
-  }
+  @if $u-align-breakpoint-toggle-v-block-middle-table__inner {
```
Comment (kllevin) [inline utilities/_u-alignments.scss]: if you change the toggle name, remember to change it here :)

File: utilities/_u-link-disguised.scss [hunk 7]
```diff
@@ -15,8 +15,15 @@
...
-$u-link-disguised-breakpoints: $default-breakpoints !default;
+/**
+ * Apply at breakpoints.
+ */
+
+// Which breakpoints you want to use?
+$u-link-disguised-breakpoints:        $default-breakpoints !default;
+
+// What do you want to apply it too?
+$u-link-disguised-breakpoint-toggle:  true !default;
```
Comment (kllevin) [inline utilities/_u-link-disguised.scss:26]: Sometimes you end the "default" state with `-base` and other times you do not. What is your thinking on the naming?

File: utilities/_u-position.scss [hunk 8]
```diff
@@ -15,8 +15,54 @@
...
+
+$u-position-breakpoint-toggle-pin-all:            false !default;
+
+$u-position-breakpoint-toggle-pin-ends:           false !default;
+
+$u-position-breakpoint-toggle-pin-sides:          false !default;
+
+$u-position-breakpoint-toggle-pin-sides-bottom:   false !default;
+
+$u-position-breakpoint-toggle-pin-sides-top:      false !default;
```
Comment (kllevin) [inline utilities/_u-position.scss:57]: Can you flip the order of `$u-position-breakpoint-toggle-pin-sides-bottom:   false !default;` and `$u-position-breakpoint-toggle-pin-sides-top:      false !default;` so they appear in the same order that they are listed lower in the code.

File: layout/_l-grid.scss [hunk 9]
```diff
@@ -38,25 +38,38 @@
  * Settings.
  */
 
-// Apply at these breakpoints
-$l-grid-breakpoints:    $default-breakpoints !default;
+/**
+ * Apply at breakpoints.
+ */
+
+// Which breakpoints you want to use?
```
Comment (kllevin) [inline layout/_l-grid.scss]: Can you change this sentence throughout to read `// Which breakpoints do you want to use?`

File: utilities/_u-columns.scss [hunk 10]
```diff
@@ -15,26 +15,51 @@
...
-$u-column-breakpoints:          $default-breakpoints !default;
+/**
+ * Apply at breakpoints.
+ */
+
+// Which breakpoints you want to use?
+$u-column-breakpoints:                        $default-breakpoints !default;
+
+// What do you want to apply it too?
+$u-column-breakpoint-toggle-base:             false !default;
```
Comment (kllevin) [inline utilities/_u-columns.scss]: Throughout you add `-base` to some, but not to all. I find the context confusing as prior to this PR base was only being used in reference to sizes, now it is used both for size and for "original" or "default" states. I would rather we only use `-base` when it is actually part of the class name we are referencing.

File: layout/_l-grid-pull.scss (hunk 3 above)
Comment (chris-pearce) [inline layout/_l-grid-pull.scss]: Good catch and yes this should be `false`, same with **Push**.

File: utilities/_u-alignments.scss (hunk 5 above)
Comment (chris-pearce) [inline utilities/_u-alignments.scss]: Yes it should, no BEM in Sass settings.

File: utilities/_u-alignments.scss (hunk 6 above)
Comment (chris-pearce) [inline utilities/_u-alignments.scss]: Done.

File: utilities/_u-columns.scss (hunk 10 above)
Comment (chris-pearce) [inline utilities/_u-columns.scss]: Good point, I'll change to `default` in these cases and keep`base` for sizing only.

File: utilities/_u-link-disguised.scss (hunk 7 above)
Comment (chris-pearce) [inline utilities/_u-link-disguised.scss:26]: I was thinking if it's just one toggle no need to say anything but maybe I should always say `default`?

Comment (chris-pearce) [inline utilities/_u-link-disguised.scss:26]: I decided if there's only 1 toggle setting then no need to say anything at the end.

File: utilities/_u-position.scss (hunk 8 above)
Comment (chris-pearce) [inline utilities/_u-position.scss:57]: Done and nice attention to detail :+1:

File: layout/_l-grid.scss (hunk 9 above)
Comment (chris-pearce) [inline layout/_l-grid.scss]: Yes.

**Comment by chris-pearce** [comment]
@kllevin I've addressed all your feedback, can you review please? See last 5 commits.

**Comment by kllevin** [comment]
LGTM thanks.

**Pull request merged**